### PR TITLE
Introducing class metadata not found

### PR DESCRIPTION
**Pull request by emulienfou**
|      Q       |   A
|------------- | -----------
| Type         | feature/improvement
| BC Break     | no
| Fixed issues | #1993 

#### Summary

* Introduce the `onClassMetadataNotFound` event along with its object
* Add support for the `onClassMetadataNotFound` event in `ResolveTargetDocumentListener`

### CODE REVIEW

File: lib/Doctrine/ODM/MongoDB/Event/OnClassMetadataNotFoundEventArgs.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+use Doctrine\Common\Persistence\Mapping\ClassMetadata;
+use Doctrine\Common\Persistence\ObjectManager;
+
+/**
+ * Class that holds event arguments for a `onClassMetadataNotFound` event.
+ *
+ * This object is mutable by design, allowing callbacks having access to it to set the
+ * found metadata in it, and therefore "cancelling" a `onClassMetadataNotFound` event
+ *
+ * @author Marco Pivetta <ocramius@gmail.com>
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Event/OnClassMetadataNotFoundEventArgs.php]: We can drop `@author` and `@since` annotation: this information is accessible using git.

File: lib/Doctrine/ODM/MongoDB/Event/OnClassMetadataNotFoundEventArgs.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+     */
+    private $foundMetadata;
+
+    /**
+     * Constructor.
+     *
+     * @param string        $className
+     * @param ObjectManager $objectManager
+     */
+    public function __construct($className, ObjectManager $objectManager)
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Event/OnClassMetadataNotFoundEventArgs.php]: This should receive a `DocumentManager` instance.

File: lib/Doctrine/ODM/MongoDB/Event/OnClassMetadataNotFoundEventArgs.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+     */
+    private $className;
+
+    /**
+     * @var ClassMetadata|null
+     */
+    private $foundMetadata;
+
+    /**
+     * Constructor.
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Event/OnClassMetadataNotFoundEventArgs.php]: This comment is useless 😉

File: lib/Doctrine/ODM/MongoDB/Event/OnClassMetadataNotFoundEventArgs.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
+<?php
+
+namespace Doctrine\ODM\MongoDB\Event;
+
+use Doctrine\Common\Persistence\Event\ManagerEventArgs;
+use Doctrine\Common\Persistence\Mapping\ClassMetadata;
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Event/OnClassMetadataNotFoundEventArgs.php]: This import should be replaced with `Doctrine\ODM\MongoDB\Mapping\ClassMetadata`.

File: lib/Doctrine/ODM/MongoDB/Tools/ResolveTargetDocumentListener.php [hunk 5]
```diff
@@ -31,6 +32,24 @@ public function addResolveTargetDocument($originalDocument, $newDocument, array
...
+    /**
+     * @param OnClassMetadataNotFoundEventArgs $args
+     *
+     * @internal this is an event callback, and should not be called directly
+     *
+     * @return void
+     */
+    public function onClassMetadataNotFound(OnClassMetadataNotFoundEventArgs $args)
+    {
+        if (array_key_exists($args->getClassName(), $this->resolveTargetDocuments)) {
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Tools/ResolveTargetDocumentListener.php]: You can use an early exit to reduce code nesting:
```php
if (! array_key_exists($args->getClassName(), $this->resolveTargetDocuments)) {
    return;
}
```

File: lib/Doctrine/ODM/MongoDB/Tools/ResolveTargetDocumentListener.php [hunk 6]
```diff
@@ -31,6 +32,24 @@ public function addResolveTargetDocument($originalDocument, $newDocument, array
...
     }
 
+    /**
+     * @param OnClassMetadataNotFoundEventArgs $args
+     *
+     * @internal this is an event callback, and should not be called directly
+     *
+     * @return void
+     */
+    public function onClassMetadataNotFound(OnClassMetadataNotFoundEventArgs $args)
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Tools/ResolveTargetDocumentListener.php]: I believe you should either move this method to #1997 (to be merged after this PR) or move the addition of the `onClassMetadataNotFound` event to `getSubscribedEvents` to this PR.

File: docs/en/reference/events.rst [hunk 7]
```diff
@@ -165,6 +165,10 @@ the life-time of their registered documents.
    loadClassMetadata - The loadClassMetadata event occurs after the
    mapping metadata for a class has been loaded from a mapping source
    (annotations/xml/yaml).
+-  onClassMetadataNotFound - Loading class metadata for a particular
```
Comment (malarzm) [inline docs/en/reference/events.rst]: On other lines the "-" is on separate line, are you sure this will render correctly?

File: tests/Doctrine/ODM/MongoDB/Tests/Mapping/ClassMetadataFactoryTest.php [hunk 8]
```diff
@@ -1,114 +1,114 @@
...
-}
-
-class TestDocument1
-{
-    private $id;
-    private $name;
-    private $other;
-    private $association;
-}
+<?php
```
Comment (malarzm) [inline tests/Doctrine/ODM/MongoDB/Tests/Mapping/ClassMetadataFactoryTest.php]: Why is this file totally changed?

File: docs/en/reference/events.rst (hunk 7 above)
Comment (emulienfou) [reply docs/en/reference/events.rst]: Just in case, I put the `-` in an other line

File: tests/Doctrine/ODM/MongoDB/Tests/Mapping/ClassMetadataFactoryTest.php (hunk 8 above)
Comment (emulienfou) [reply tests/Doctrine/ODM/MongoDB/Tests/Mapping/ClassMetadataFactoryTest.php]: Oups! I revert a commit by it replace the CRLF by LF

File: lib/Doctrine/ODM/MongoDB/Event/OnClassMetadataNotFoundEventArgs.php [hunk 9]
```diff
@@ -0,0 +1,63 @@
+<?php
+
+namespace Doctrine\ODM\MongoDB\Event;
+
+use Doctrine\Common\Persistence\Event\ManagerEventArgs;
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Event/OnClassMetadataNotFoundEventArgs.php]: You should remove this import: we've got out own `ManagerEventArgs` class that you should extend. This will also fix the test error.

File: lib/Doctrine/ODM/MongoDB/Tools/ResolveTargetDocumentListener.php [hunk 10]
```diff
@@ -2,21 +2,34 @@
...
+use Doctrine\ODM\MongoDB\Events;
 use Doctrine\ODM\MongoDB\Mapping\ClassMetadata;
 
 /**
  * ResolveTargetDocumentListener
  *
  * Mechanism to overwrite document interfaces or classes specified as association targets.
  */
-class ResolveTargetDocumentListener
+class ResolveTargetDocumentListener implements EventSubscriber
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Tools/ResolveTargetDocumentListener.php]: I believe we should move these changes to #1997 entirely. This PR should only add the `onClassMetadataNotFound` event, with any other functionality based on it being introduced in #1997. As it is, this duplicates a change from #1997 that is not necessary at all.

File: tests/Doctrine/ODM/MongoDB/Tests/Mapping/ClassMetadataFactoryTest.php (hunk 8 above)
Comment (alcaeus) [inline tests/Doctrine/ODM/MongoDB/Tests/Mapping/ClassMetadataFactoryTest.php]: Please revert all changes to this file: it is impossible to review changes to this file. If there is a refactoring that needs to be done here, please extract it to a separate PR. Thanks!

Comment (alcaeus) [reply tests/Doctrine/ODM/MongoDB/Tests/Mapping/ClassMetadataFactoryTest.php]: Changes reverted 👍

**Comment by alcaeus** [comment]
After a short chat with @emulienfou, I've rebased the PR and removed unrelated changes (e.g. ClassMetadataFactoryTest). I've added a small test to ensure that the metadata factory returns the metadata instance set in the listener.

**Comment by alcaeus** [comment]
Build against lowest dependency versions seem to fail, I'll have to debug that some more. Maybe it's time to bump the lowest version of doctrine/common 🙄

**Comment by alcaeus** [comment]
Thanks @emulienfou! 🎉
